### PR TITLE
WIP: Add Freesurfer 7 

### PR DESCRIPTION
**Pull request by Shotgunosine**
I added Freesurfer 7 and some additional jobs on circleci to test and deploy it. Circlci config will likely need to be fixed before this is complete.

### CODE REVIEW

**Comment by Shotgunosine** [comment]
@PeerHerholz and @andysworth this resolves #57 with the potential exception of the replacement recon-all script. I'd be happy to get your feedback on it.

**Comment by PeerHerholz** [comment]
Ahoi hoi @Shotgunosine,

thx for the fantastic work and effort. The same to you @andysworth.
I checked the code and tests, LGTM.  The plan re the `recon-all` script would be include it once we have it, eh?

**Comment by Shotgunosine** [comment]
Yeah, That's what I'm thinking.

**Comment by PeerHerholz** [comment]
Awesome! @andysworth, do you want to have a look as well? After that or otherwise, we could go ahead and merge!?

**Comment by AndysWorth** [comment]
Sorry for the slow response.  This looks good to me.  I just noticed that I got a response on the Freesurfer mailing list from Doug Greve:

> Hi Andy, this is a bug introduced in 7.1.1, some kind of race condition.
> You can fix it by commenting out this line in recon-all
>   if($OMP_NUM_THREADS > 1 && ! $LHonly && ! $RHonly) set cmd = ($cmd --parallel) # just hemis
> doug

Thanks for your strong work @Shotgunosine!

**Comment by PeerHerholz** [comment]
Cool, thanks @AndysWorth! @Shotgunosine should we go ahead and merge? WDYT re the `recon-all` script adaptation above?

**Comment by Shotgunosine** [comment]
We should add a sed command to the docker command, probably the best place
would be in neurodocker. @AndysWorth, do you think you could open an issue
there and submit a PR?

**Comment by AndysWorth** [comment]
Will do!

**Comment by PeerHerholz** [comment]
Hi folks,

just checking: we can merge this, right? @Shotgunosine @AndysWorth

**Comment by AndysWorth** [comment]
Sorry, I was looking forward to making that contribution but it seems that I'm slowing this down.  It looks like I won't have time until I'm off work for the holidays.  @Shotgunosine if you want to go ahead with your suggestion in the meanwhile, that would be great.  I know it's a small change but I don't feel confident to just do it without taking the time to make sure I fully understand what's going on.

**Comment by Shotgunosine** [comment]
I might not be able to get to it until around the same time period. Maybe @llevitis would be able to take a look

**Comment by AndysWorth** [comment]
I've installed the fix here at Flywheel -- once I have had a chance to run it a few times I'll submit it to neurodocker.

**Comment by Shotgunosine** [comment]
@AndysWorth have you had a chance to send the fix over to neurodocker?

**Comment by AndysWorth** [comment]
@Shotgunosine Sorry, no 😞 .  There are two places that need patching in recon-all and the fixes have worked well at Flywheel so it's about time!  I'll work on making the time.  Thanks for the reminder.

**Comment by Shotgunosine** [comment]
@AndysWorth, sounds good. It'd be great if you're able to make the changes this week, I've got a bunch of sessions I need to run and I'd love to be able to use this.

**Comment by AndysWorth** [comment]
Okay @Shotgunosine, I will!

**Comment by AndysWorth** [comment]
Hey @Shotgunosine , I (finally) managed to patch Freesurfer 7.7.1 in neurodocker.  See [neurodocker PR#384](https://github.com/ReproNim/neurodocker/pull/384).  I did this exact same change to the official Freesurfer docker image version of 7.1.1 and it has been run successfully many times on various Flywheel instances.  I _have not_ tested this new neurodocker version, which is for a slightly different version of centos (6 vs. 7).  Is this PR sufficient for you to give it a try?  All it should do is prevent crashing.

**Comment by AndysWorth** [comment]
@Shotgunosine I've created a repository [recon-all-v7.1.1-parallel-patch](https://github.com/AndysWorth/recon-all-v7.1.1-parallel-patch).  Right now it just has the patch file that can be installed with
`patch /usr/local/freesurfer/bin/recon-all parallel.patch`
Is this enough?  If not, what else should I do?

**Comment by Shotgunosine** [comment]
@AndysWorth, can you update the neurodocker PR to pull that repo run the bash script?

**Comment by AndysWorth** [comment]
@Shotgunosine I'm not familiar enough with neurodocker.  Where would I do that?

**Comment by AndysWorth** [comment]
@Shotgunosine It sounds like they don't want the patch in neurodocker so the following should go somewhere around [here](https://github.com/Shotgunosine/freesurfer/blob/fs_7/generate_freesurfer_images.sh#L96)?

```
curl -LJO https://github.com/AndysWorth/recon-all-v7.1.1-parallel-patch/tarball/master
tar zxf AndysWorth-recon-all-v7.1.1-parallel-patch-e5ff2fe.tar.gz
cd AndysWorth-recon-all-v7.1.1-parallel-patch-e5ff2fe
patch $FREESURFER_HOME/bin/recon-all parallel.patch 
patch $FREESURFER_HOME/bin/quantifyThalamicNuclei.sh quantifyThalamicNuclei.patch
cd ..
rm -rf AndysWorth-recon-all-v7.1.1-parallel-patch-e5ff2fe
```

**Comment by Shotgunosine** [comment]
I’ll see if I can use the script you wrote to get a command into
neurodocker that’ll make them happy.

**Comment by Shotgunosine** [comment]
With the new clarifications can you take care of it @Andysworth?

**Comment by bpinsard** [comment]
Hi, are there any blockers to merging that very old PR? 
It would be great to have latest versions integrated into repronim/containers. 
Anything I can do to help?
Should we restart a PR from https://github.com/e-dads/freesurfer ?
Thanks!

**Comment by Shotgunosine** [comment]
I haven't looked at this in a while, but I think this convo with Satra is where things were left off:

https://github.com/ReproNim/neurodocker/pull/384#issuecomment-812187770

@bpinsard, are you comfortable with the structure of everything to just make that change? It'd take me a while to reestablish context here.

**Comment by Shotgunosine** [comment]
Actually, looks like we could just jump to Freesurfer 7.4.1. Would that be alright?

**Comment by bpinsard** [comment]
I am not sure I have played enough with neurodocker to fully grasp the issue to be fixed. 
If I understand, that would mean: 
- inserting the patch in your branch as @satra suggested into that file: https://github.com/Shotgunosine/freesurfer/blob/fs_7/generate_freesurfer_images.sh 
- regenerating Dockerfile and Apptainer recipes.

The latest version would be great! I don't think there is really an interest in redeploying each older version.

**Comment by Remi-Gau** [comment]
I can have a look to try to fix some of the merge conflicts on the circle-CI config. I have tried to "standardize" across bids apps to make the maintenance easier.

**Comment by Shotgunosine** [comment]
I'm working on 7.4.1 now, to do that I've got to update the neurodocker version, which breaks some of the syntax. I'm working on it though.

**Comment by bpinsard** [comment]
I had the [generate_freesurfer_images.sh](https://github.com/Shotgunosine/freesurfer/blob/fs_7/generate_freesurfer_images.sh) running with the following changes:
https://github.com/courtois-neuromod/freesurfer/tree/pr/67  (requires neurodocker master) 
If I understand correctly the patches are not needed with that version, right?

**Comment by Shotgunosine** [comment]
@Remi-Gau I took a first pass at resolving merge conflicts on the circleci config, but I've been out of the circleci game for a minute. Do you think you can fix the config now, or have I just made a mess of things?

**Comment by Remi-Gau** [comment]
had a quick look
it seems to make sense but may need to go for a deeper dive: those diffs are hard to read

**Comment by Remi-Gau** [comment]
also seems that some file changes are showing in the PR when they are already in the default branch, no?

**Comment by Shotgunosine** [comment]
I tried to rebase the PR thinking that that would be more straightfoward, but may have just made a mess. Happy to do a force rollback if you think that'd be easier.

**Comment by Remi-Gau** [comment]
yeah if you can try to get a cleaner diff it'd be great

let me know if I can help

**Comment by Shotgunosine** [comment]
Alright, should be cleaner now.

**Comment by Remi-Gau** [comment]
yup indeed

File: .circleci/config.yml [hunk 1]
```diff
@@ -171,6 +171,99 @@ jobs:
...
+        no_output_timeout: 6h
+    - run:
+        name: group2 test 2
+        command: |
+          docker run -ti --rm --read-only -v ~/license/license.txt:/license.txt --tmpfs /tmp --tmpfs /var/tmp -v ${HOME}/data/ds114_test2:/bids_dataset -v ${HOME}/data/ds114_test2_freesurfer_precomp_v6.0.0:/outputs bids/${CIRCLE_PROJECT_REPONAME}_fs7 /bids_dataset /outputs group2 --license_file=/license.txt  && mkdir -p ${HOME}/outputs2/ && sudo mv ${HOME}/data/ds114_test2_freesurfer_precomp_v6.0.0/00_group* ${HOME}/outputs2/ && cat ${HOME}/outputs2/00_group2_stats_tables/lh.aparc.thickness.tsv && cat ${HOME}/outputs2/00_group2_stats_tables/euler.tsv
```
Comment (Remi-Gau) [inline .circleci/config.yml]: do you mind if we "multiline" those commands?

I find docker calls super hard to parse in general so I try to stick to one argument, parameter per line.

See here for example: https://github.com/bids-apps/freesurfer/blob/6b76873f65dd5c027c54e3440bd1ae9423e462f6/.circleci/config.yml#L102-L115

**Review by Remi-Gau — DISMISSED** [review]

File: .circleci/config.yml (hunk 1 above)
Comment (Shotgunosine) [reply .circleci/config.yml]: sure, I'll clean those up.

**Comment by Remi-Gau** [comment]
also just thinking: if we agree that we want both freesurfer 6 and 7 to be available it should probably be mentioned in the readme how to pull each from docker hub.
Does not have to be part of this PR: can be done once a release has been done and a tagged image has been pushed

File: .circleci/config.yml (hunk 1 above)
Comment (Remi-Gau) [reply .circleci/config.yml]: thanks

**Comment by Shotgunosine** [comment]
I think we will want both freesurfer 6 and 7 available. HCP still uses freesurfer 6.

**Comment by Remi-Gau** [comment]
having a look locally at the whole config.

it looks very much like this is doing exactly the same thing for both images, no?

like literally running the same commands with both containers

sounds like this should probably be refactored, no?

**Comment by Shotgunosine** [comment]
Definitely, do you think you could refactor it? It should be straightforward, I just don't know circle's syntax very well.

**Comment by Remi-Gau** [comment]
> Definitely, do you think you could refactor it? It should be straightforward, I just don't know circle's syntax very well.

yup I will give it a go

**Comment by Remi-Gau** [comment]
ok refactored things now let's see how this baby flies on the circle-ci side...

**Comment by Remi-Gau** [comment]
Note that given that one of the preivous build did not complete because we hit the circle CI "wall time" I think we may have to move to github actions that will give us 6 hours before we get cancelled. I am not as good with caching things in GHA action though...

https://app.circleci.com/pipelines/github/bids-apps/freesurfer/82/workflows/bc88a5ed-4c03-4c1e-ba7a-dd0b33577e60/jobs/340

But let's wait and see how things are going on this build... Not holding my breath.

**Comment by Remi-Gau** [comment]
note that the image is using an antiquated bids-validator version (0.19.8 when the current one is 1.12.0): may be worth bumping this up.

**Comment by Remi-Gau** [comment]
ok the image build is failing because of some conda hanging: https://app.circleci.com/pipelines/github/bids-apps/freesurfer/87/workflows/d6fe415d-5abd-427d-95c9-792ab1bf0073/jobs/355

trying locally

**Comment by Remi-Gau** [comment]
The only that changes in the bash script used to generate the different recipe is the freesurfer version, no?

I think this should be refactored too

**Comment by Remi-Gau** [comment]
@Shotgunosine can you check if you can build the image using the docker recipe: it seems to fail in CI and for me locally.

May need to adapt the neurodocker command

**Comment by Shotgunosine** [comment]
If hangs forever solving the environment for me locally as well. I'm trying to see if there's a way to install pandas that doesn't hang, but I haven't found anything yet. Might need to add mamba to neurodocker.

**Comment by Remi-Gau** [comment]
> If hangs forever solving the environment for me locally as well. I'm trying to see if there's a way to install pandas that doesn't hang, but I haven't found anything yet. Might need to add mamba to neurodocker.

maybe specifying the exact version of each package may help?

**Comment by Shotgunosine** [comment]
I was able to get it to build with a local version of neurodocker that I added a mamba option to, and I've submitted a pull request for that: https://github.com/ReproNim/neurodocker/pull/562

edit: PR merged, I've updated the neurodocker version in the bash script accordingly.

**Comment by Shotgunosine** [comment]
@Remi-Gau I've got no idea why circle is trying to do anything with /home/circleci/license/license.txt (which is the [source of the failure](https://app.circleci.com/pipelines/github/bids-apps/freesurfer/93/workflows/7c7de9e6-0a06-41f3-a6d1-710c8e49873c/jobs/378?invite=true#step-103-251664_78)). As far as I can tell I've changed all references to that directory and the build ran past it when I changed the entry point to python. Do you have any ideas? I'm stumped.

Edit: Nevermind , figured it out.

**Comment by Remi-Gau** [comment]
> Edit: Nevermind , figured it out.

🚀

File: .circleci/config.yml [hunk 2]
```diff
@@ -60,51 +70,65 @@ jobs:
...
+            -v /tmp/workspace/license_fs<< parameters.version >>/license.txt:/license.txt \
             -v /tmp/workspace/data/ds114_test1:/bids_dataset \
-            -v ${HOME}/outputs1:/outputs bids/${CIRCLE_PROJECT_REPONAME} \
-            /bids_dataset /outputs participant \
-              --participant_label 01 \
-              --license_file=/license.txt --stages autorecon1
+            -v ${HOME}/outputs1:/outputs \
+            --tmpfs /tmp --tmpfs /var/tmp \
+              bids/${CIRCLE_PROJECT_REPONAME}_fs<< parameters.version >> \
+                /bids_dataset
```
Comment (Remi-Gau) [inline .circleci/config.yml]:
```suggestion
                /bids_dataset \
```

**Comment by Remi-Gau** [comment]
will make a dummy change to retrigger circle ci

**Comment by Remi-Gau** [comment]
getting somse validator error: will bump its version

**Comment by Remi-Gau** [comment]
> getting somse validator error: will bump its version

totally missed that you had done that and completely refactored the bash script: you've been busy

**Comment by Shotgunosine** [comment]
I had to bump the node version to get one that was available for ubuntu-jammy, so I went ahead and bumped the bids validator version to the latest. I had only bumped node to 16.x and the validator says it needs at least 18.x so I've bumped node to 18.x and we'll try again.

**Comment by Shotgunosine** [comment]
Bah, looks like circleci webhooks are broken at the moment. Is there anyone who can trigger a manual build?

**Comment by Remi-Gau** [comment]
> Is there anyone who can trigger a manual build?

just did: let's see how this one goes

**Comment by Remi-Gau** [comment]
see if we can create specific dependencies between the steps of the circle ci so that test for fs6 starts when the build for fs6 is done.

Currently the tests can only start when both builds are done.

image

**Comment by Shotgunosine** [comment]
> see if we can create specific dependencies between the steps of the circle ci so that test for fs6 starts when the build for fs6 is done.
> 
> Currently the tests can only start when both builds are done.
> 
> image

Yep, found the [docs on matrix](https://circleci.com/docs/configuration-reference/#matrix) and they explain how to set it up. The circleci graph makes it looks like the dependencies are correct now.

File: Dockerfile
Comment (Remi-Gau) [inline Dockerfile:1]: shouldn't we just delete the Dockerfile?

**Comment by Remi-Gau** [comment]
> Yep, found the [docs on matrix](https://circleci.com/docs/configuration-reference/#matrix) and they explain how to set it up. The circleci graph makes it looks like the dependencies are correct now.

nice!!!

thanks for looking into this

**Comment by Shotgunosine** [comment]
@Remi-Gau, the tests are failing because they can't find anything at [/home/circleci/data/ds114_test1_freesurfer_precomp_v6.0.0/00_group*](https://app.circleci.com/pipelines/github/bids-apps/freesurfer/103/workflows/f4ee1e9a-63ab-4913-b0a6-149f25573870/jobs/419?invite=true#step-109-3496_112). That part of the circleconfig predates this PR. Do you know what's going on with that?

**Comment by Remi-Gau** [comment]
get the same bug locally:

```bash
# get download script
wget https://raw.githubusercontent.com/bids-apps/maintenance-tools/main/utils/get_data_from_osf.sh

# get gata into home dir from https://osf.io/9q7dv/
bash get_data_from_osf.sh ds114_test1_freesurfer_precomp_v6.0.0 

# create target dir
mkdir -p ${HOME}/outputs1 ${HOME}/outputs2

# try to copy
sudo mv ${HOME}/data/ds114_test1_freesurfer_precomp_v6.0.0/00_group* ${HOME}/outputs1/
```
no group folder

```bash
$ ls ${HOME}/data/ds114_test1_freesurfer_precomp_v6.0.0/       
  
fsaverage  lh.EC_average  rh.EC_average  sub-01  sub-02
```

let me dig into the history of the repo to see what was happening in the good old days

**Comment by Remi-Gau** [comment]
here is the circle ci "step" that was used before I started working in the circle-ci:

https://github.com/bids-apps/freesurfer/blob/8c364a62ffbdef3b3e4f81a1c6bf4de5e9c2b5cd/.circleci/config.yml#L66C67-L66C67

more readable

```bash
docker run -ti --rm --read-only \
  -v ~/license/license.txt:/license.txt \
  --tmpfs /tmp --tmpfs /var/tmp \
  -v ${HOME}/data/ds114_test1:/bids_dataset \
  -v ${HOME}/data/ds114_test1_freesurfer_precomp_v6.0.0:/outputs \
    bids/${CIRCLE_PROJECT_REPONAME} /bids_dataset /outputs group2 \
      --license_file=/license.txt && \
mkdir -p ${HOME}/outputs1/ && \
sudo mv ${HOME}/data/ds114_test1_freesurfer_precomp_v6.0.0/00_group* ${HOME}/outputs1/ && \
cat ${HOME}/outputs1/00_group2_stats_tables/lh.aparc.thickness.tsv && \
cat ${HOME}/outputs1/00_group2_stats_tables/euler.tsv
```

**Comment by Remi-Gau** [comment]
I think I it is "just" a path issue: trying to fix in https://github.com/bids-apps/freesurfer/pull/67/commits/765a747844da7745a49b19c1087a8098ce213589

**Comment by Shotgunosine** [comment]
@Remi-Gau, it passed! Is there anything else we should fix for this PR?

**Comment by Remi-Gau** [comment]
woohoo!!!

I will merge and make a new release: this should trigger a push to docherhub.

**Comment by bpinsard** [comment]
Wow, glad my ping triggered a new release! Thanks for all the hard work! Waiting on the dockerhub push and will make a PR to repronim, can't wait to try it out!

**Comment by Remi-Gau** [comment]
> Wow, glad my ping triggered a new release! Thanks for all the hard work! Waiting on the dockerhub push and will make a PR to repronim, can't wait to try it out!

FYI it is already on dockerhub but it is a bit of a mess at the moment:

see here
https://github.com/bids-apps/freesurfer/issues/75#issuecomment-1699911795

any thoughts on what would make most sense for users?
